### ISSUE TITLE
dlib link error iOS

### ISSUE THREAD

**Issue by dmwesterhoff** [zweigraf/face-landmarking-ios#7]
Hi, wondering if you can speak or provide guidance to this specific link time issue I'm having

_USER_ERROR__missing_dlib_all_source_cpp_file__OR__inconsistent_use_of_DEBUG_or_ENABLE_ASSERTS_preprocessor_directives

This happened when I tried to port the libdlib.a and wrapper classes to my project- I have attempted to change all build settings in my project to look like the example project. Any ideas?

**Comment by dmwesterhoff** [comment]
Believe to have tracked down the issue, it was due to a cocoapods flag that was using DEBUG=1 as a preprocessor macro

**Comment by willard-yuan** [comment]
When I compiles the project, I also face the problem. Anyone knows how to deal with the problem?

```sh
ld: warning: ignoring file /Users/willard/codes/swift/dlib_app/face-landmarking-ios/DisplayLiveSamples/lib/libdlib.a, missing required architecture i386 in file /Users/willard/codes/swift/dlib_app/face-landmarking-ios/DisplayLiveSamples/lib/libdlib.a (2 slices)
Undefined symbols for architecture i386:
  "_USER_ERROR__missing_dlib_all_source_cpp_file__OR__inconsistent_use_of_DEBUG_or_ENABLE_ASSERTS_preprocessor_directives_", referenced from:
      __GLOBAL__sub_I_DlibWrapper.mm in DlibWrapper.o
ld: symbol(s) not found for architecture i386
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by zweigraf** [comment]
Don't use the simulator :) The camera won't work anyway. This error just means that the Dlib library is not compiled for the simulator (i386).

**Comment by dboyliao** [comment]
I encounter the same issue when I build `dlib` from source and drag the static library into my xcode project. I finally manage to fix this error and I think it is good to have a step by step approach here for those who are not familier with compilation/linking. See [here](http://stackoverflow.com/questions/34591254/how-to-build-dlib-for-ios) for instructions of building `dlib` for `iOS`.

The origin of this linking error comes from `dlib/threads/threads_kernel_shared.h`. It's good to have a look on the comments in it. To fix the bug, do as following:
1. Go to `Build Settings > Apple LLVM - Custom Compiler Flags > Other C Flags` and enter `-DDLIB_JPEG_SUPPORT -DDLIB_NO_GUI_SUPPORT -DNDEBUG -DDLIB_USE_BLAS -DDLIB_USE_LAPACK`
  - these are flags which defines some macros for the preprocessor and will include different features in the compiled `dlib` static library.
2. Once you have `libdlib.a` compiled, drag it into your xcode project.
3. Go to the `Build Settings` in your xcode project and search for `Preprocessor Macros`.
  - As @zweigraf said in the [README.md](https://github.com/zweigraf/face-landmarking-ios/blob/master/README.md#credits), make sure you use the same macros as the c flags you use for compiling `libdlib.a`.
  - As an example here, you should set `Preprocessor Macros` as `DLIB_JPEG_SUPPORT DLIB_NO_GUI_SUPPORT NDEBUG DLIB_USE_BLAS DLIB_USE_LAPACK` if you follow step 1.
  - Note that the `-D` is removed for each flags.
4. Then you have it! You should be able to compile and link successfully.

**Comment by scm-ns** [comment]
For me what ended up solving the issue was build Dlib under Release mode, by specifying it in the scheme. Thanks a lot for all the help!

**Comment by torrtuga** [comment]
@dboyliao Hi,
I did add for both Custom Compiler Flag as well as Preprocessor Macros.
But am still getting the same `"_USER_ERROR__missing_dlib_all_source_cpp_file__OR__inconsistent_use_of_DEBUG_or_ENABLE_ASSERTS_preprocessor_directives_", referenced from:` error.

I just added one line : `#include <dlib/image_processing.h>` to check for dlib. And this line gave me the error.

**Comment by kennybatista** [comment]
The error still continued in my case, until I realized that I had made a silly mistake. I had added the preprocessor macros but only on debug mode, I fixed it by making sure that it was inputed on both Debug and Release.
![screen shot 2017-08-28 at 2 31 31 pm](https://user-images.githubusercontent.com/10367842/29794634-99697b82-8bfd-11e7-9897-7aab1bcdd542.png)

**Comment by GLinnik21** [comment]
Same here

**Comment by PavanMore** [comment]
I am facing same issue tried all issues above but I have stuck for flag -std=c++11, when I set it in build setting in project I get the more errors and I cannot build library without that flag. So its like blocker , I am trying to build 19.4 version. Any help will be appreciated.

**Comment by WimVriend** [comment]
Same here: that nasty bugger did not want to go...

Finally I realised that it is also possible to use dlib without using cmake, as mentioned under chapter [Compiling C++ Examples Without CMake](http://dlib.net/compile.html). I added _/all/source.cpp_ to my project and removed libdlib.a. Of course, I also had to add the correct path to the Header Search path (1 for dlib itself and one for libjpeg)

Hope this helps anyone!
